contracts/BookMarketplace.sol

```diff
@@ -44,15 +48,98 @@ contract BookMarketplace is ERC721URIStorage {
         bool sold
     );
 
+    event ListingUpdated(
+        uint256 indexed tokenId,
+        address seller,
+        address owner,
+        uint256 instantPrice,
+        uint256 startingPrice,
+        bool allowBid,
+        Bid[] bidList,
+        bool sold
+    );
+
+    /* modifiers */
+    
+    
+    // Modifier prevents function from being called by anyone other than the market owner
+    modifier onlyMarketOwner() {
+        require(
+            marketOwner == msg.sender,
+            "Only the marketplace owner can call this function."
+        );
+        _;
+    }
+
+    /* constructor */
+
     constructor() ERC721("BookExchange Tokens", "BOOK") {
-        owner = payable(msg.sender);
+        marketOwner = payable(msg.sender);
+    }
+
+    /* functions */
+
+    function getMarketOwnerAddress() public view returns(address) {
+        return marketOwner;
+    }
+
+    // change the price of a listing
+    function updateInstantPrice(uint256 tokenId, uint256 newPrice) public payable onlyMarketOwner {
+        Listing storage listing = idToMarketItem[tokenId];
+        listing.instantPrice = newPrice;
+        idToMarketItem[tokenId] = listing;
+    }
+
+    // get the price of a listing
+    function getInstantPrice(uint256 tokenId) public view returns (uint256) {
+        Listing storage listing = idToMarketItem[tokenId];
+        return listing.instantPrice;
+    }
+
+    // Adds bids to a bid list by tokenId
+    function addBid(uint256 tokenId, uint256 bidAmount) public payable {
+        Listing storage listing = idToMarketItem[tokenId];
+        require(
+            listing.allowBid == true,
+            "This item does not allow bids."
+        );
+        require(
+            listing.sold == false,
+            "This item has been sold."
+        );
+        require(
+            bidAmount >= listing.instantPrice,
+            "Bid must be greater than or equal to the current price."
+        );
+        require(
+            msg.value == bidAmount,
+            "Bid amount must be equal to the amount sent." 
+        );
+        Bid memory newBid;
+        newBid.bidder = msg.sender;
+        newBid.bidAmount = bidAmount;
+        listing.bidList.push(newBid);
```

Comment:
What if the same bidder makes another bid on the same listing? Do we push that or do we modify the original bid?

Reply:
Keeping both would allow us to keep a history of bids, which was my thoughts on it. Though we could modify the original.

Reply:
@MichaelFauht what do you think?

Reply:
I think that the discussion about whether or not we should have people prepay bids should come first, if we do, then we should modify the original bid and get them to pay the difference, if not, then just add a bid instead. (otherwise we will have to refund their last bid as they submit the new one or something like that).

Reply:
It would make sense to keep track of all the bids instead of updating the original. That would allow for transparency with the price and expose someone trying to drive it up artificially

Reply:
Good stuff @MichaelFauht, and I think with the bidding fees introduced, we could cover the cost of memories on Blockchain

Reply:
Same as the previous, see updated commits when pushed.